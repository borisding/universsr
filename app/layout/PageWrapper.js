import React from 'react';
import Helmet from 'react-helmet';
import PropTypes from 'prop-types';

const PageWrapper = ({ title, children }) => (
  <>
    <Helmet title={title} />
    {children}
  </>
);

PageWrapper.propTypes = {
  children: PropTypes.node.isRequired,
  title: PropTypes.string.isRequired
};

export default PageWrapper;
