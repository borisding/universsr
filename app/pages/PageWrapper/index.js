import React from 'react';
import Helmet from 'react-helmet';
import PropTypes from 'prop-types';

export default function PageWrapper({ title, children }) {
  return (
    <>
      <Helmet title={title} />
      {children}
    </>
  );
}

PageWrapper.propTypes = {
  children: PropTypes.node.isRequired,
  title: PropTypes.string.isRequired
};
