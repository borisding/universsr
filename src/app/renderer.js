import React from 'react';
import DocumentTitle from 'react-document-title';
import serialize from 'serialize-javascript';
import flushChunks from 'webpack-flush-chunks';
import { renderToString } from 'react-dom/server';
import { flushChunkNames } from 'react-universal-component/server';
import { matchRoutes, renderRoutes } from 'react-router-config';
import { StaticRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import storeFactory from './store';
import routes from './routes';

// preload data for matched route
function prefetchBranchData(store, url) {
  try {
    const branch = matchRoutes(routes, url);
    const promises = branch.map(({ route, match }) => {
      const { loadData } = route;
      const { dispatch } = store;

      if (match && match.isExact && loadData) {
        return Array.isArray(loadData)
          ? Promise.all(loadData.map(action => dispatch(action(match))))
          : dispatch(loadData(match));
      }

      return Promise.resolve(null);
    });

    return Promise.all(promises);
  } catch (err) {
    throw err;
  }
}

// export default server renderer and receiving stats
// also, should allow it to be mounted as middleware for production usage
export default function serverRenderer({ clientStats }) {
  return async (req, res, next) => {
    try {
      const context = {};
      const store = storeFactory();

      await prefetchBranchData(store, req.url);

      const appString = renderToString(
        <Provider store={store}>
          <StaticRouter location={req.url} context={context}>
            {renderRoutes(routes)}
          </StaticRouter>
        </Provider>
      );

      const pageTitle = DocumentTitle.rewind();
      const chunksOptions = { chunkNames: flushChunkNames() };
      const { scripts, stylesheets } = flushChunks(clientStats, chunksOptions);

      const preloadedState = serialize(store.getState(), { isJSON: true });
      const { statusCode = 200, redirectUrl } = context;
      const { nonce } = res.locals;

      if ([301, 302].includes(statusCode) && redirectUrl) {
        return res.redirect(statusCode, redirectUrl);
      }

      res.status(statusCode);
      res.render('index', {
        pageTitle,
        appString,
        preloadedState,
        stylesheets,
        scripts,
        nonce
      });
    } catch (err) {
      next(new Error(err));
    }
  };
}
