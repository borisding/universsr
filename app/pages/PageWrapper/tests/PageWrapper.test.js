import React from 'react';
import renderer from 'react-test-renderer';
import PageWrapper from '../';

describe('PageWrapper', () => {
  it('renders correctly', () => {
    const props = { children: ['<div>page child</div>'], title: 'Page Title' };
    const tree = renderer.create(<PageWrapper {...props} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
