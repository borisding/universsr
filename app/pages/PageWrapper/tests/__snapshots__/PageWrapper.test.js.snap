// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageWrapper renders correctly 1`] = `"<div>page child</div>"`;
